perun-core/src/main/java/cz/metacentrum/perun/core/impl/Utils.java

```diff
@@ -158,8 +163,8 @@ public static String join(Iterable<?> collection, String separator) {
 	 * @return List<UserExtSource> all additional ExtSources from the subject, returned list will never contain null value
 	 * @throws InternalErrorException
 	 */
-	public static List<UserExtSource> extractAdditionalUserExtSources(PerunSession sess, Map<String, String> subjectFromExtSource) {
-		List<UserExtSource> additionalUserExtSources = new ArrayList<>();
+	public static List<RichUserExtSource> extractAdditionalUserExtSources(PerunSession sess, Map<String, String> subjectFromExtSource) {
```

Comment:
This whole method should be refactored to smaller parts with better Javadoc. There should be info what is the expected format of additionalUserExtSource in the subject, which part is which and how it will be parsed.

Reply:
Ok, I've splitted the method to smaller parts and added some comments and javadocs.